app/src/ui/create-branch/create-branch-dialog.tsx

```diff
@@ -101,9 +106,53 @@ export class CreateBranch extends React.Component<
   ICreateBranchProps,
   ICreateBranchState
 > {
+  private api?: API
+
+  private checkBranchRules = debounce(async (branchName: string) => {
+    if (
+      this.api === undefined ||
+      this.props.upstreamGitHubRepository === null ||
+      branchName === '' ||
+      this.state.currentError !== null
+    ) {
+      return
+    }
+
+    const branchRules = await this.api.fetchRepoRulesForBranch(
+      this.props.upstreamGitHubRepository.owner.login,
+      this.props.upstreamGitHubRepository.name,
+      branchName
+    )
+
+    if (branchRules.length === 0) {
+      return
+    }
+
+    let errMsg: string | null = null
+    const parsedRules = parseRepoRules(branchRules)
+
+    if (parsedRules.creationRestricted) {
+      errMsg = `Branch name '${branchName}' is restricted by repo rules`
+    } else if (parsedRules.branchNamePatterns.hasRules) {
+      const errors = parsedRules.branchNamePatterns.getFailedRules(branchName)
+
+      if (errors.length > 0) {
+        errMsg = `Branch name '${branchName}' may be restricted by repo rules: ${errors.join('. ')}`
+      }
+    }
+
+    if (errMsg) {
+      this.setState({ currentError: { error: new Error(errMsg), isBlocking: false }})
+    }
+  }, 500)
+
   public constructor(props: ICreateBranchProps) {
     super(props)
 
+    if (props.upstreamGitHubRepository !== null && props.gitHubAccount !== null) {
+      this.api = API.fromAccount(props.gitHubAccount)
+    }
```

Comment:
API objects are not expensive to create, rather than setting it up in the constructor and thereby missing any prop updates for the account we should just instantiate it in the debounced method.

Reply:
Handling this now in https://github.com/desktop/desktop/pull/16904/commits/d9e3c23a2bcefc6d1abcc96cbc3420b80e5325b2, thanks! I looked at moving this and the accounts change into `componentWillReceiveUpdates` so it only has to be computed when they change, but I'm not a React expert and React's docs say not to use that method for new code, so I just left it here for now.